internal/batch/batch_processor.go

```diff
@@ -411,76 +445,106 @@ func (bp *batchProcessor) maskContext(ctx context.Context, msg *fftypes.Message,
 			continue
 		}
-		for _, topic := range msg.Header.Topics {
-			pinString, contextOrPin, err := bp.maskContext(ctx, msg, topic)
+		var pins fftypes.FFStringArray
+		if isPrivate {
+			pins = make(fftypes.FFStringArray, len(msg.Header.Topics))
+			state.msgPins[*msg.Header.ID] = pins
+		}
+		for i, topic := range msg.Header.Topics {
+			pinString, contextOrPin, err := bp.maskContext(ctx, state, msg, topic)
 			if err != nil {
 				return nil, err
 			}
 			contextsOrPins = append(contextsOrPins, contextOrPin)
-			if msg.Header.Group != nil {
-				msg.Pins = append(msg.Pins, pinString /* contains the nonce as well as the pin hash */)
-				pinsAssigned = true
+			if isPrivate {
+				pins[i] = pinString
 			}
 		}
-		if pinsAssigned {
-			// It's important we update the message pins at this phase, as we have "spent" a nonce
-			// on this topic from the database. So this message has grabbed a slot in our queue.
-			// If we fail the dispatch, and redo the batch sealing process, we must not allocate
-			// a second nonce to it (and as such modifiy the batch payload).
-			err := bp.database.UpdateMessage(ctx, msg.Header.ID,
-				database.MessageQueryFactory.NewUpdate(ctx).Set("pins", msg.Pins),
-			)
-			if err != nil {
-				return nil, err
+	}
+	return contextsOrPins, bp.flushNonceState(ctx, state)
+}
+
+func (bp *batchProcessor) flushNonceState(ctx context.Context, state *DispatchState) error {
+
+	// Flush all the assigned nonces to the DB
+	for hash, nonce := range state.noncesAssigned {
+		dbNonce := &fftypes.Nonce{
+			Hash:  &hash,
+			Nonce: nonce,
+		}
+		if nonce == 0 {
+			if err := bp.database.InsertNonce(ctx, dbNonce); err != nil {
+				return err
+			}
+		} else {
+			if err := bp.database.UpdateNonce(ctx, dbNonce); err != nil {
+				return err
+			}
+		}
+	}
+
+	// It's important we update the message pins at this phase, as we have "spent" a nonce
+	// on this topic from the database. So this message has grabbed a slot in our queue.
+	// If we fail the dispatch, and redo the batch sealing process, we must not allocate
+	// a second nonce to it (and as such modifiy the batch payload).
+	//
+	// Note to consider database transactions, we do not update the msg.Pins array, or the
+	// message cache, until after the Retry/RunAsGroup has ended in sealBatch.
+	for _, msg := range state.Messages {
+		if pins, ok := state.msgPins[*msg.Header.ID]; ok {
+			if err := bp.database.UpdateMessage(ctx, msg.Header.ID,
+				database.MessageQueryFactory.NewUpdate(ctx).Set("pins", pins),
+			); err != nil {
+				return err
 			}
 		}
 	}
-	return contextsOrPins, nil
+
+	return nil
 }
 
 func (bp *batchProcessor) sealBatch(state *DispatchState) (err error) {
 	err = bp.retry.Do(bp.ctx, "batch persist", func(attempt int) (retry bool, err error) {
 		return true, bp.database.RunAsGroup(bp.ctx, func(ctx context.Context) (err error) {
+			state.noncesAssigned = make(map[fftypes.Bytes32]int64)
+			state.msgPins = make(map[fftypes.UUID]fftypes.FFStringArray)
```

Comment:
Are these specifically (re)initialized at this moment (rather than in `initFlushState` with the rest of the struct) in order to clear state upon retries? Might be worth a comment.

Reply:
Yes - so we re-read Nonces from the DB. I'll add a comment